consul_exporter.go

```diff
@@ -0,0 +1,235 @@
+// as Prometheus metrics. It implements prometheus.Collector.
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	services := make(chan []*consul_api.ServiceEntry)
+
+	go e.queryClient(services)
+
+	e.mutex.Lock() // To protect metrics from concurrent collects.
+	defer e.mutex.Unlock()
+
+	// Reset metrics.
+	e.serviceNodesTotal.Reset()
+	e.serviceNodesHealthy.Reset()
+
+	e.setMetrics(services)
+
+	ch <- e.up
+	ch <- e.clusterServers
+	ch <- e.nodeCount
+	ch <- e.serviceCount
+
+	e.serviceNodesTotal.Collect(ch)
+	e.serviceNodesHealthy.Collect(ch)
+}
+
+func (e *Exporter) queryClient(services chan<- []*consul_api.ServiceEntry) {
+	defer close(services)
+
+	// How many peers are in the Consul cluster?
+	peers, err := e.client.Status().Peers()
+
+	if err != nil {
+		e.up.Set(0)
+		log.Printf("Query error is %v", err)
+		return
+	}
+
+	// We'll use peers to decide that we're up.
+	e.up.Set(1)
+	e.clusterServers.Set(float64(len(peers)))
+
+	// How many nodes are registered?
+	nodes, _, err := e.client.Catalog().Nodes(&consul_api.QueryOptions{})
+
+	if err != nil {
+		// FIXME: How should we handle a partial failure like this?
+	} else {
+		e.nodeCount.Set(float64(len(nodes)))
+	}
+
+	// Query for the full list of services.
+	serviceNames, _, err := e.client.Catalog().Services(&consul_api.QueryOptions{})
+	e.serviceCount.Set(float64(len(serviceNames)))
+
+	if err != nil {
+		// FIXME: How should we handle a partial failure like this?
+		return
+	}
+
+	e.serviceCount.Set(float64(len(serviceNames)))
+
+	for s := range serviceNames {
+		s_entries, _, err := e.client.Health().Service(s, "", false, &consul_api.QueryOptions{})
+
+		if err != nil {
+			log.Printf("Failed to query service health: %v", err)
+			continue
+		}
+
+		services <- s_entries
+	}
+}
+
+func (e *Exporter) setMetrics(services <-chan []*consul_api.ServiceEntry) {
+	// Each service will be an array of ServiceEntry structs.
+	for service := range services {
+		if len(service) == 0 {
+			// Not sure this should ever happen, but catch it just in case...
+			continue
+		}
+
+		// We should have one ServiceEntry per node, so use that for total nodes.
+		e.serviceNodesTotal.WithLabelValues(service[0].Service.Service).Set(float64(len(service)))
+
+		for _, entry := range service {
+			// We have a Node, a Service, and one or more Checks. Our
+			// service-node combo is passing if all checks have a `status`
+			// of "passing."
+
+			passing := 1
+
+			for _, hc := range entry.Checks {
+				if hc.Status != consul.HealthPassing {
+					passing = 0
+					break
+				}
+			}
+
+			log.Printf("%v/%v status is %v", entry.Service.Service, entry.Node.Node, passing)
+
+			e.serviceNodesHealthy.WithLabelValues(entry.Service.Service, entry.Node.Node).Set(float64(passing))
+		}
+	}
+}
+
+func main() {
+	var (
+		listenAddress = flag.String("web.listen-address", ":9105", "Address to listen on for web interface and telemetry.")
```

Comment:
Oops, this port has been taken by two new exporters in the meantime. Could you change it to 9107? (https://github.com/prometheus/prometheus/wiki/Default-port-allocations)